src/Components/Components/src/Sections/SectionOutlet.cs

```diff
@@ -31,20 +31,20 @@ Task IComponent.SetParametersAsync(ParameterView parameters)
     {
         parameters.SetParameterProperties(this);
 
-        if (string.IsNullOrEmpty(Name))
+        if (SectionId is null)
         {
-            throw new InvalidOperationException($"{GetType()} requires a non-empty string parameter '{nameof(Name)}'.");
+            throw new InvalidOperationException($"{GetType()} requires a non-empty string parameter '{nameof(SectionId)}'.");
         }
 
-        if (Name != _subscribedName)
+        if (SectionId != _subscribedSectionId)
```

Comment:
Very good point!

Indeed this is becoming too complex. We decided to use object `SectionId` instead of string `Name` to avoid any naming issues. But, as you pointed out, the users can assign anything to `SectionId` (could be reference type in your example or could be string which is immutable and there will be equality). So now  if we decided (most likely not) to leave object `SectionId` as it is then we are relying on our users to know how reference and value types are allocated in the memory. With the current solution we cannot force them to use only object for `SectionId`. That could potentially introduce confusion for the users. We need something else. 
I like the way you think to fix equality checking but it is becoming too complex.

I think we should go back to string `Name`. It is simple and intuitive. The only downside to this is when there are two `SectionOutlet`s rendered on the same page with the same `Name` but an exception will be thrown saying that `SectionOutlet` with the same name already exists, so this is fine. Or another case when there is a `SectionContent` with some `Name` and it happens that there already a `SectionOutlet` with the same `Name` on this page. The user will see that the content is rendered and will figure out that the `SectionOutlet` had the same `Name`, so this is fine too.

What do you think?

Reply:
> 2. You want to use value equality for a reference type
Let's say I've defined a custom MySectionIdentifier class that implements IEquatable<MySectionIdentifier>, and I plan on using an instance of this class as the SectionId for my SectionOutlet. There would currently be no way of enabling a SectionContent to render to my SectionOutlet without somehow obtaining a reference to the exact MySectionIdentifier passed to my SectionOutlet.

I thought that was the whole point to use the exact same object for `SectionId` of `SectionContent` as in `SectionOutlet`. The only way is to pass it as parameter to the child component or use static field in the parent component.

This brings me to a solution to fix the issue with casting object. We can introduce new class `SectionIdentifier` and make `SectionId` of type `SectionIdentifier`. But I think it is better to go back to string `Name`.

Reply:
As discussed, we are implementing the design where `SectionId` is object